File: backend-node/src/services/rateLimitService.js
/**
 * Rate limiting service using sliding window algorithm.
 *
 * Implements 10 requests per minute per user using:
 * - Redis (preferred for production, multi-instance deployments)
 * - In-memory fallback (single-instance only, resets on restart)
 */

// Rate limit configuration
const RATE_LIMIT_REQUESTS = 10; // Max requests
const RATE_LIMIT_WINDOW_SECONDS = 60; // Per minute

// Redis connection settings
const REDIS_URL = process.env.REDIS_URL || null;

// Global Redis client (lazy initialized)
let redisClient = null;
let redisAvailable = null;

/**
 * Lazy-initialize Redis client. Returns null if Redis unavailable.
 */
async function getRedisClient() {
  if (redisAvailable === false) {
    return null;
  }

  if (redisClient !== null) {
    return redisClient;
  }

  if (!REDIS_URL) {
    redisAvailable = false;
    console.log('REDIS_URL not configured, using in-memory rate limiting');
    return null;
  }

  try {
    const Redis = require('ioredis');
    redisClient = new Redis(REDIS_URL);
    await redisClient.ping();
    redisAvailable = true;
    console.log('Redis rate limiting initialized');
    return redisClient;
  } catch (error) {
    redisAvailable = false;
    console.warn(`Redis unavailable, falling back to in-memory: ${error.message}`);
    return null;
  }
}

/**
 * In-memory rate limiter using sliding window.
 * WARNING: Only suitable for single-instance deployments.
 */
class InMemoryRateLimiter {
  constructor() {
    this.requests = new Map(); // userId -> array of timestamps
  }

  /**
   * Check if user is within rate limit.
   * @param {number} userId - User ID
   * @returns {[boolean, number, number]} Tuple of (allowed, remaining, resetInSeconds)
   */
  checkRateLimit(userId) {
    const now = Date.now() / 1000;
    const windowStart = now - RATE_LIMIT_WINDOW_SECONDS;

    // Get user's requests
    let userRequests = this.requests.get(userId) || [];

    // Remove expired timestamps
    userRequests = userRequests.filter(ts => ts > windowStart);

    const currentCount = userRequests.length;
    const remaining = Math.max(0, RATE_LIMIT_REQUESTS - currentCount);

    if (currentCount >= RATE_LIMIT_REQUESTS) {
      // Calculate when oldest request will expire
      const oldest = userRequests.length > 0 ? Math.min(...userRequests) : now;
      const resetIn = Math.ceil(oldest + RATE_LIMIT_WINDOW_SECONDS - now) + 1;
      this.requests.set(userId, userRequests);
      return [false, 0, resetIn];
    }

    // Record this request
    userRequests.push(now);
    this.requests.set(userId, userRequests);
    return [true, remaining - 1, RATE_LIMIT_WINDOW_SECONDS];
  }
}

// Singleton in-memory rate limiter
const inMemoryLimiter = new InMemoryRateLimiter();

/**
 * Redis-based rate limiter using sorted sets for sliding window.
 */
class RedisRateLimiter {
  constructor(client) {
    this.client = client;
  }

  /**
   * Check if user is within rate limit using Redis.
   * @param {number} userId - User ID
   * @returns {Promise<[boolean, number, number]>} Tuple of (allowed, remaining, resetInSeconds)
   */
  async checkRateLimit(userId) {
    const key = `rate_limit:${userId}`;
    const now = Date.now() / 1000;
    const windowStart = now - RATE_LIMIT_WINDOW_SECONDS;

    // Execute atomically using multi
    const multi = this.client.multi();
    
    // Remove old entries outside the window
    multi.zremrangebyscore(key, '-inf', windowStart);
    // Count current requests in window
    multi.zcard(key);
    // Get oldest timestamp
    multi.zrange(key, 0, 0, 'WITHSCORES');

    const results = await multi.exec();
    const currentCount = results[1][1];

    if (currentCount >= RATE_LIMIT_REQUESTS) {
      const oldestData = results[2][1];
      let resetAt = RATE_LIMIT_WINDOW_SECONDS;
      if (oldestData && oldestData.length >= 2) {
        resetAt = Math.ceil(parseFloat(oldestData[1]) + RATE_LIMIT_WINDOW_SECONDS - now);
      }
      return [false, 0, Math.max(1, resetAt)];
    }

    // Add current request
    const uniqueKey = `${now}-${Math.random()}`;
    await this.client.zadd(key, now, uniqueKey);
    await this.client.expire(key, RATE_LIMIT_WINDOW_SECONDS + 10);

    const remaining = RATE_LIMIT_REQUESTS - currentCount - 1;
    return [true, remaining, RATE_LIMIT_WINDOW_SECONDS];
  }
}

/**
 * Check if a user is within their rate limit.
 * @param {number} userId - User ID
 * @returns {Promise<[boolean, number, number]>} Tuple of (allowed, remaining, resetInSeconds)
 */
async function checkRateLimit(userId) {
  const redis = await getRedisClient();
  
  if (redis) {
    const limiter = new RedisRateLimiter(redis);
    return limiter.checkRateLimit(userId);
  }
  
  return inMemoryLimiter.checkRateLimit(userId);
}

/**
 * Get rate limit headers for HTTP response.
 * @param {number} remaining - Remaining requests
 * @param {number} resetIn - Seconds until reset
 * @returns {object} Headers object
 */
function getRateLimitHeaders(remaining, resetIn) {
  return {
    'X-RateLimit-Limit': RATE_LIMIT_REQUESTS.toString(),
    'X-RateLimit-Remaining': Math.max(0, remaining).toString(),
    'X-RateLimit-Reset': Math.ceil(Date.now() / 1000 + resetIn).toString(),
  };
}

module.exports = {
  RATE_LIMIT_REQUESTS,
  RATE_LIMIT_WINDOW_SECONDS,
  checkRateLimit,
  getRateLimitHeaders,
};

File: backend/app/services/rate_limit_service.py
"""
Rate limiting service using sliding window algorithm.

Implements 10 requests per minute per user using:
- Redis (preferred for production, multi-instance deployments)
- In-memory fallback (single-instance only, resets on restart)

Design decisions:
- Sliding window provides smoother rate limiting than fixed windows
- Redis atomic operations prevent race conditions
- In-memory uses threading locks for thread safety
- Rate limit state is ephemeral (not persisted to DB) - acceptable for rate limits
"""
import time
import os
from typing import Tuple, Optional
from collections import defaultdict
import threading
import logging

logger = logging.getLogger(__name__)

# Rate limit configuration
RATE_LIMIT_REQUESTS = 10  # Max requests
RATE_LIMIT_WINDOW_SECONDS = 60  # Per minute

# Redis connection settings
REDIS_URL = os.getenv("REDIS_URL", None)

# Global Redis client (lazy initialized)
_redis_client = None
_redis_available = None


def _get_redis_client():
    """
    Lazy-initialize Redis client. Returns None if Redis unavailable.
    Caches the result to avoid repeated connection attempts.
    """
    global _redis_client, _redis_available
    
    if _redis_available is False:
        return None
    
    if _redis_client is not None:
        return _redis_client
    
    if not REDIS_URL:
        _redis_available = False
        logger.info("REDIS_URL not configured, using in-memory rate limiting")
        return None
    
    try:
        import redis
        _redis_client = redis.from_url(REDIS_URL, decode_responses=True)
        _redis_client.ping()  # Test connection
        _redis_available = True
        logger.info("Redis rate limiting initialized")
        return _redis_client
    except Exception as e:
        _redis_available = False
        logger.warning(f"Redis unavailable, falling back to in-memory: {e}")
        return None


class InMemoryRateLimiter:
    """
    Thread-safe in-memory rate limiter using sliding window.
    
    WARNING: Only suitable for single-instance deployments.
    State is lost on restart and not shared between instances.
    """
    
    def __init__(self):
        # Dict of user_id -> list of request timestamps
        self._requests: dict[int, list[float]] = defaultdict(list)
        self._lock = threading.Lock()
    
    def check_rate_limit(self, user_id: int) -> Tuple[bool, int, int]:
        """
        Check if user is within rate limit.
        
        Returns:
            Tuple of (allowed: bool, remaining: int, reset_in_seconds: int)
        """
        now = time.time()
        window_start = now - RATE_LIMIT_WINDOW_SECONDS
        
        with self._lock:
            # Remove expired timestamps (outside current window)
            self._requests[user_id] = [
                ts for ts in self._requests[user_id] 
                if ts > window_start
            ]
            
            current_count = len(self._requests[user_id])
            remaining = max(0, RATE_LIMIT_REQUESTS - current_count)
            
            if current_count >= RATE_LIMIT_REQUESTS:
                # Calculate when oldest request will expire
                oldest = min(self._requests[user_id]) if self._requests[user_id] else now
                reset_in = int(oldest + RATE_LIMIT_WINDOW_SECONDS - now) + 1
                return False, 0, reset_in
            
            # Record this request
            self._requests[user_id].append(now)
            return True, remaining - 1, RATE_LIMIT_WINDOW_SECONDS


class RedisRateLimiter:
    """
    Redis-based rate limiter using sorted sets for sliding window.
    
    Uses atomic Lua script to prevent race conditions.
    Suitable for multi-instance production deployments.
    """
    
    # Lua script for atomic rate limit check and increment
    # This runs atomically on Redis, preventing race conditions
    RATE_LIMIT_SCRIPT = """
    local key = KEYS[1]
    local now = tonumber(ARGV[1])
    local window = tonumber(ARGV[2])
    local limit = tonumber(ARGV[3])
    local window_start = now - window
    
    -- Remove old entries outside the window
    redis.call('ZREMRANGEBYSCORE', key, '-inf', window_start)
    
    -- Count current requests in window
    local current = redis.call('ZCARD', key)
    
    if current >= limit then
        -- Get oldest timestamp to calculate reset time
        local oldest = redis.call('ZRANGE', key, 0, 0, 'WITHSCORES')
        local reset_at = 0
        if oldest[2] then
            reset_at = oldest[2] + window - now
        end
        return {0, 0, math.ceil(reset_at)}
    end
    
    -- Add current request
    redis.call('ZADD', key, now, now .. '-' .. math.random())
    
    -- Set TTL slightly longer than window to auto-cleanup
    redis.call('EXPIRE', key, window + 10)
    
    local remaining = limit - current - 1
    return {1, remaining, window}
    """
    
    def __init__(self, redis_client):
        self._redis = redis_client
        self._script = self._redis.register_script(self.RATE_LIMIT_SCRIPT)
    
    def check_rate_limit(self, user_id: int) -> Tuple[bool, int, int]:
        """
        Check if user is within rate limit using Redis.
        
        Returns:
            Tuple of (allowed: bool, remaining: int, reset_in_seconds: int)
        """
        key = f"rate_limit:chat:{user_id}"
        now = time.time()
        
        try:
            result = self._script(
                keys=[key],
                args=[now, RATE_LIMIT_WINDOW_SECONDS, RATE_LIMIT_REQUESTS]
            )
            allowed, remaining, reset_in = result
            return bool(allowed), int(remaining), int(reset_in)
        except Exception as e:
            logger.error(f"Redis rate limit error: {e}")
            # Fail open on Redis errors - rely on quota limits as backup
            return True, RATE_LIMIT_REQUESTS, RATE_LIMIT_WINDOW_SECONDS


# Singleton instances
_in_memory_limiter = InMemoryRateLimiter()
_redis_limiter: Optional[RedisRateLimiter] = None


def check_rate_limit(user_id: int) -> Tuple[bool, int, int]:
    """
    Check if user is within rate limit.
    
    Automatically uses Redis if available, falls back to in-memory.
    
    Args:
        user_id: The user's ID
        
    Returns:
        Tuple of (allowed: bool, remaining: int, reset_in_seconds: int)
    """
    global _redis_limiter
    
    redis_client = _get_redis_client()
    
    if redis_client:
        if _redis_limiter is None:
            _redis_limiter = RedisRateLimiter(redis_client)
        return _redis_limiter.check_rate_limit(user_id)
    
    return _in_memory_limiter.check_rate_limit(user_id)


def get_rate_limit_headers(remaining: int, reset_in: int) -> dict:
    """
    Generate standard rate limit headers for HTTP response.
    """
    return {
        "X-RateLimit-Limit": str(RATE_LIMIT_REQUESTS),
        "X-RateLimit-Remaining": str(remaining),
        "X-RateLimit-Reset": str(int(time.time()) + reset_in),
    }
